### PR TITLE
Bugfix/retransmission

### PR DESCRIPTION
**Pull request by faderskd**
Fixes https://github.com/allegro/hermes/issues/978

### CODE REVIEW

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/BrokersClusterService.java [hunk 1]
```diff
@@ -58,4 +72,34 @@ public boolean topicExists(Topic topic) {
...
+            return numberOfAssignmentsForConsumersGroups(consumerGroupsForSubscriptions) == requiredTotalNumberOfAssignments;
+        } catch (Exception e) {
+            logger.error("Failed to check assignments for topic " + topic.getQualifiedName() + " subscriptions", e);
+            return false;
+        }
+    }
+
+    private int numberOfAssignmentsForConsumersGroups(List<String> consumerGroupsIds) throws ExecutionException, InterruptedException {
+        Collection<ConsumerGroupDescription> consumerGroupsDescriptions = adminClient.describeConsumerGroups(consumerGroupsIds).all().get().values();
+        Collection<MemberDescription> memberDescriptions = consumerGroupsDescriptions.stream().flatMap(desc -> desc.members().stream()).collect(Collectors.toList());
```
Comment (spooz) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/BrokersClusterService.java]: we actually dont need to collect that, can be left as stream, minor

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java [hunk 2]
```diff
@@ -35,11 +41,26 @@ public TopicContentTypeMigrationService(SubscriptionRepository subscriptionRepos
 
     public void notifySubscriptions(Topic topic, Instant beforeMigrationInstant) {
         waitUntilOffsetsAvailableOnAllKafkaTopics(topic, CHECK_OFFSETS_AVAILABLE_TIMEOUT);
-        subscriptionRepository.listSubscriptionNames(topic.getName()).forEach(s ->
-                notifySingleSubscription(topic, beforeMigrationInstant, s)
+        logger.info("Offsets available on all partitions of topic {}", topic.getQualifiedName());
+        subscriptionRepository.listSubscriptions(topic.getName())
```
Comment (spooz) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java]: can be a private method, used 2 times

File: hermes-test-helper/src/main/java/pl/allegro/tech/hermes/test/helper/endpoint/JerseyClientFactory.java [hunk 3]
```diff
@@ -15,8 +15,8 @@ public static Client create() {
     public static ClientConfig createConfig() {
         ClientConfig clientConfig = new ClientConfig();
         clientConfig.property(ClientProperties.ASYNC_THREADPOOL_SIZE, 10);
-        clientConfig.property(ClientProperties.CONNECT_TIMEOUT, 5000);
-        clientConfig.property(ClientProperties.READ_TIMEOUT, 20000);
+        clientConfig.property(ClientProperties.CONNECT_TIMEOUT, 50000);
+        clientConfig.property(ClientProperties.READ_TIMEOUT, 200000);
```
Comment (cristaloleg) [inline hermes-test-helper/src/main/java/pl/allegro/tech/hermes/test/helper/endpoint/JerseyClientFactory.java]: ~1 minute & ~3 minutes respectively?

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java [hunk 4]
```diff
@@ -210,7 +211,12 @@ public void updateTopic(TopicName topicName, PatchData patch, String modifiedBy)
                 );
             }
             topicRepository.updateTopic(modified);
+
             if (!retrieved.wasMigratedFromJsonType() && modified.wasMigratedFromJsonType()) {
+                logger.info("Waiting until all subscriptions have consumers assigned during topic {} content type migration...", topicName.qualifiedName());
+                topicContentTypeMigrationService.waitUntilAllSubscriptionsHasConsumersAssigned(modified,
+                        Duration.ofSeconds(topicProperties.getCheckTopicSubscriptionsAssignmentCompletedTimeoutSeconds()));
```
Comment (cristaloleg) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java]: woah `getCheckTopicSubscriptionsAssignmentCompletedTimeoutSeconds`

Comment (faderskd) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java]: done

File: hermes-test-helper/src/main/java/pl/allegro/tech/hermes/test/helper/endpoint/JerseyClientFactory.java (hunk 3 above)
Comment (faderskd) [reply hermes-test-helper/src/main/java/pl/allegro/tech/hermes/test/helper/endpoint/JerseyClientFactory.java]: done

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java (hunk 2 above)
Comment (faderskd) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java]: done

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/BrokersClusterService.java (hunk 1 above)
Comment (faderskd) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/kafka/service/BrokersClusterService.java]: done

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java [hunk 5]
```diff
@@ -17,8 +21,9 @@
 
     private static final Logger logger = LoggerFactory.getLogger(TopicContentTypeMigrationService.class);
 
-    public static final Duration CHECK_OFFSETS_AVAILABLE_TIMEOUT = Duration.ofSeconds(1);
-    public static final Duration INTERVAL_BETWEEN_OFFSETS_AVAILABLE_CHECK_RETRIES = Duration.ofMillis(500);
+    private static final Duration CHECK_OFFSETS_AVAILABLE_TIMEOUT = Duration.ofSeconds(30);
+    private static final Duration INTERVAL_BETWEEN_OFFSETS_AVAILABLE_CHECK = Duration.ofMillis(500);
+    private static final Duration INTERVAL_BETWEEN_ASSIGNMENTS_COMPLETED_CHECK = Duration.ofMillis(500);
```
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java]: can we have these 3 params configurable?

Comment (faderskd) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java]: Currently we have not any configuration for that, if something goes wrong user can perform retransmission one more time. If we will need this in future we will make different pull request :)

Comment (dankraw) [reply hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicContentTypeMigrationService.java]: 👍

**Pull request merged**